### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by cmazakas**

### CODE REVIEW

File: include/sleip/dynamic_array.hpp [hunk 1]
```diff
@@ -157,9 +131,25 @@ struct dynamic_array : boost::empty_value<Allocator, 0>
...
+
+    auto* const p = boost::to_address(d.get());
+
+    boost::alloc_construct_n(alloc_, p, count);
+
+    data_ = d.release();
+    size_ = count;
+  }
+
+  explicit dynamic_array(default_init_t, size_type count, Allocator const& alloc = Allocator())
```
Comment (K-ballo) [inline include/sleip/dynamic_array.hpp]: Tag type should go after count, to be consistent with constructor taking a `T const& value`.

File: include/sleip/dynamic_array.hpp [hunk 2]
```diff
@@ -569,21 +568,7 @@ struct dynamic_array : boost::empty_value<Allocator, 0>
...
-      data_[idx * unroll_size + 2] = value;
-      data_[idx * unroll_size + 3] = value;
-      data_[idx * unroll_size + 4] = value;
-      data_[idx * unroll_size + 5] = value;
-      data_[idx * unroll_size + 6] = value;
-      data_[idx * unroll_size + 7] = value;
-    }
-
-    for (size_type idx = num_rolls * unroll_size; idx < size_; ++idx) { data_[idx] = value; }
+    for (size_type idx = 0; idx < size(); ++idx) { data_[idx] = value; }
```
Comment (K-ballo) [inline include/sleip/dynamic_array.hpp]: As long as you are already pulling `<algorithm>`, might just as well call `std::fill`.

File: include/sleip/dynamic_array.hpp [hunk 3]
```diff
@@ -612,6 +599,7 @@ template <class T, class Allocator>
 auto
 operator==(dynamic_array<T, Allocator> const& lhs, dynamic_array<T, Allocator> const& rhs) -> bool
 {
+  if (lhs.size() != rhs.size()) { return false; }
```
Comment (K-ballo) [inline include/sleip/dynamic_array.hpp]: This check is already performed by `std::equal` for random access iterators.

File: include/sleip/dynamic_array_fwd.hpp [hunk 4]
```diff
@@ -1,37 +1,33 @@
 #ifndef SLEIP_DYNAMIC_ARRAY_FWD_HPP_
 #define SLEIP_DYNAMIC_ARRAY_FWD_HPP_
 
-#include <boost/core/noinit_adaptor.hpp>
+#if __has_include(<vcruntime.h>)
+#include <vcruntime.h>
```
Comment (K-ballo) [inline include/sleip/dynamic_array_fwd.hpp:7]: Is this for `_HAS_CXX17`?

File: include/sleip/dynamic_array_fwd.hpp [hunk 5]
```diff
@@ -1,37 +1,33 @@
...
+#endif
 
-#if __has_include(<memory_resource>)
-#include <memory_resource>
-#define SLEIP_HAS_PMR 1
-#else
-#define SLEIP_HAS_PMR 0
+#include <boost/config.hpp>
+
+#if !__has_include(<memory_resource>) || (!BOOST_MSVC &&__cplusplus < 201703L) || (BOOST_MSVC && !_HAS_CXX17)
```
Comment (K-ballo) [inline include/sleip/dynamic_array_fwd.hpp]: Nit: space around `&&`

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -32,7 +33,7 @@ sleip_add_test(operations)
 sleip_add_test(fancy_pointer)
 target_link_libraries(fancy_pointer PRIVATE Threads::Threads)
 if (UNIX AND NOT APPLE)
-  target_link_libraries(fancy_pointer PRIVATE rt)
+  target_link_libraries(fancy_pointer PRIVATE Threads::Threads rt)
```
Comment (K-ballo) [inline test/CMakeLists.txt:37]: Does "apple" not need `Threads::Threads`?

File: test/noinit.cpp [hunk 7]
```diff
@@ -14,7 +14,7 @@ namespace pmr = boost::container::pmr;
 void
 test_default_allocator()
 {
```
Comment (K-ballo) [inline test/noinit.cpp:22]: This whole test is odd...

File: include/sleip/dynamic_array_fwd.hpp (hunk 4 above)
Comment (cmazakas) [reply include/sleip/dynamic_array_fwd.hpp:7]: Indeed! I'll comment there.

File: test/CMakeLists.txt (hunk 6 above)
Comment (cmazakas) [reply test/CMakeLists.txt:37]: This was a weird change. I re-compiled using clang w/o sanitizers and I got an error about missing DSO symbols. I'm not sure how CMake does it link flags but it seems like just linking against librt messes things up.

In the ideal case, this is only a Linux problem. I think OS X has its own thing whereas Linux requires linking against librt

File: test/noinit.cpp (hunk 7 above)
Comment (cmazakas) [reply test/noinit.cpp:22]: This should hopefully be fixed now!

**Pull request merged**